src/utils/manager.py

```diff
@@ -185,55 +186,128 @@ def mount(
+def umount(mountpoint: Union[str, os.PathLike]) -> None:
     """Unmount an NFS share.
 
     Args:
         mountpoint: NFS share mountpoint to unmount.
-        force: Pass `--force` option to umount. Default: False.
 
     Raises:
         Error: Raised if NFS share umount operation fails.
     """
-    cmd = ["umount"]
-    if force:
-        cmd.append("--force")
-    cmd.append(mountpoint)
+
+    def unlink_file(path: Union[str, os.PathLike]) -> None:
+        """Deletes the file in `path` if it exists.
+
+        Raises:
+            Error: Raised if the file could not be deleted.
+        """
+        try:
+            pathlib.Path(path).unlink(missing_ok=True)
+        except OSError as e:
+            _logger.error(f"Failed to delete file {path}. Reason:\n{e}")
+            raise Error(f"Failed to unmount {mountpoint}")
+
+    _logger.debug(f"Unmounting NFS share at mountpoint {mountpoint}")
+    autofs_id = _mountpoint_to_autofs_id(mountpoint)
+
+    unlink_file(f"/etc/auto.{autofs_id}")
+    unlink_file(f"/etc/auto.master.d/{autofs_id}.autofs")
+
     try:
-        _logger.debug(f"Unmounting NFS share at mountpoint {mountpoint}")
-        subprocess.run(cmd, stderr=subprocess.PIPE, check=True, text=True, timeout=120)
-        shutil.rmtree(mountpoint, ignore_errors=True)
-    except subprocess.CalledProcessError as e:
-        _logger.error(f"{e} Reason:\n{e.stderr}")
-        raise Error(f"Failed to unmount {mountpoint}")
-    except subprocess.TimeoutExpired as e:
-        _logger.error(f"{e} Reason:\n{e.stderr}")
-        raise Error(f"Failed to unmount {mountpoint} failed after {e.timeout:.0f}s")
+        systemd.service_reload("autofs", restart_on_failure=True)
+    except systemd.SystemdError as e:
+        _logger.error(f"Failed to unmount {mountpoint}. Reason:\n{e}")
+        raise Error(f"Failed to unmount {mountpoint}.")
+
+    shutil.rmtree(mountpoint, ignore_errors=True)
+
+
+def _trigger_autofs() -> None:
+    """Triggers a mount on all filesystems handled by autofs.
+
+    This function is useful to make autofs-managed mounts appear on the
+    `/proc/mount` file, since they could be unmounted when reading the file.
+    """
+    for fs in _mounts("autofs"):
+        _logger.info(f"triggering automount for `{fs.mountpoint}`")
+        try:
+            os.scandir(fs.mountpoint).close()
+        except OSError as e:
+            # Not critical since it could also be caused by unrelated mounts,
+            # but should be good to log it in case this causes problems.
+            _logger.warning(f"Could not trigger automount for `{fs.mountpoint}`. Reason:\n{e}")
+
+
+def _mountpoint_to_autofs_id(mountpoint: Union[str, os.PathLike]) -> str:
+    """Converts a mountpoint to its corresponding autofs id.
+
+    Args:
+        mountpoint: NFS share mountpoint.
+    """
+    path = pathlib.Path(mountpoint).resolve()
+    return str(path).lstrip("/").replace("/", "-")
+
+
+def _mounts(fstype_prefix: str) -> Iterator[MountInfo]:
```

Comment:
> This type annotation is incorrect. yield returns a generator function that behaves like an iterator, but it's a generator, not an iterator. Each iteration over a generator is a new function call. So each time you iterate over _mounts(...), you're performing a new look up in /proc/mounts.

Yes, that's the intention of this code. This generator just abstracts the actions of "load /proc/mounts file" and "iterate through the mounts" in a function call instead of doing it manually every time. Also, it's not really useful to mark this as a `Generator[MountInfo]` because we don't use "generator exclusive" functionality. Marking it as returning something that adheres to the `Iterator` interface allows us to change the internals of the function as long as the returned object is something that can be iterated. 

> I'm not convinced that this generator is necessary. Generators are useful when you have to perform an operation on a dynamic variable each iteration (like grabbing the position of a widget), but here we're really just grabbing a list of specific filesystems from the /proc/mount file. This doesn't require a generator as we can just punch the static values into a list.

Generators are also useful when you want to lazily load the output of an IO operation (such as reading a file) without having to load the whole file at once. The previous implementation did exactly this, which is a bit wasteful because the OS already loads the whole file in memory.

Reply:
> Generators are also useful when you want to lazily load the output of an IO operation such as reading a file without having to load the whole file at once. The previous implementation did exactly this, which is a bit wasteful because the OS already loads the whole file in memory. 

Heh - I don't love the original implementation either, and I'm the one who wrote it :sweat_smile: 

You have the right idea for why you would use the `Iterator` annotation, but that's not totally how type annotations work in Python. They're supposed to indicate the return type of the function, so while `Generator` implements `Iterator`, the return type is still a `Generator`. Type checkers like mypy would start complaining; we don't use mypy because type annotations are very much bolted onto Python and if static typing is super important, I would argue that we should be using a different language such as Rust or Go.

I think we can heavily simplify this implementation (we're not running on memory starved systems :sweat_smile:) and still use the regex. We'd just add another regex for autofs:

```python3
_nfs_mount_parser = re.compile(
    r"""
        (?P<endpoint>\S+?)\s+
        (?P<mountpoint>\S+?)\s
        (?P<fstype>nfs\S*?)\s
        (?P<options>\S+?)\s
        (?P<freq>\d+?)\s+
        (?P<passno>\d+)
    """,
    re.VERBOSE,
)
_autofs_mount_parser = re.compile(
    r"""
        (?P<endpoint>\S+?)\s+
        (?P<mountpoint>\S+?)\s
        (?P<fstype>autofs\S*?)\s
        (?P<options>\S+?)\s
        (?P<freq>\d+?)\s+
        (?P<passno>\d+)
    """,
    re.VERBOSE,
)
```

__Then in the `mounts` function:__

```python3
def _mounts(fstype: Literal["nfs", "autofs"]) -> List[MountInfo]:
    dispatch = {"nfs": _nfs_mount_parser, "autofs": _autofs_mount_parser}
    mount = pathlib.Path("/proc/mounts").read_text()
    # Don't need to worry about a `KeyError` here since _mounts is a private function
    # and we control the inputs. The statement below finds all statements that
    # match the procompiled regex.
    return [MountInfo(**m.groupdict()) for m in dispatch[fstype].finditer(mounts)]
```

^ This approach removes the added complexity of using a generator, and fixes the previous implementation where we weren't processing the whole file at once and rather line-by-line. I appreciate lazy loading, but I don't think we need it here specifically. If we were processing a large file (say a couple gigabytes), then I'd advocate for using a buffer here, but we don't need it for `/proc/mounts`. We'd really want a buffer and/or generator if we were scanning the entire file system for mounts rather than just using _/proc/mounts_ directly.

Reply:
> You have the right idea for why you would use the `Iterator` annotation, but that's not totally how type annotations work in Python. They're supposed to indicate the return type of the function, so while `Generator` implements `Iterator`, the return type is still a `Generator`. Type checkers like mypy would start complaining; we don't use mypy because type annotations are very much bolted onto Python and if static typing is super important, I would argue that we should be using a different language such as Rust or Go.

That's not true? mypy is [perfectly capable of understanding structural subtyping](
https://mypy.readthedocs.io/en/stable/protocols.html#predefined-protocols-reference), and even the mypy documentation [recommends using `Iterator` instead of generator](https://mypy.readthedocs.io/en/stable/kinds_of_types.html#generators):

> A basic generator that only yields values can be succinctly annotated as having a return type of either [Iterator[YieldType]](https://docs.python.org/3/library/typing.html#typing.Iterator) or [Iterable[YieldType]](https://docs.python.org/3/library/typing.html#typing.Iterable)

And mypy perfectly knows when you cast a `Generator` into an `Iterator`:
https://mypy-play.net/?mypy=latest&python=3.12&gist=073673a3751c9eb79ffd6e6dad968a05

> I think we can heavily simplify this implementation (we're not running on memory starved systems 😅) and still use the regex. We'd just add another regex for autofs:

I still think having to maintain two regexps that essentially do the same thing is maintainance effort that we don't really need to carry. Python code is perfectly readable for processing a file line by line, and a regexp has very cryptic syntax that we need to mentally parse each and every time we want to do a modification.

> ^ This approach removes the added complexity of using a generator, and fixes the previous implementation where we weren't processing the whole file at once and rather line-by-line.

Leaving the "buffering" argument aside which I can agree that we really don't need, I don't see where's the complexity of using a generator? It allows using clean code that you can tell what it does at a glance, and it has the added bonus of being lazy and avoiding doing more work than it needs to.

Reply:
> [recommends using Iterator instead of generator](https://mypy.readthedocs.io/en/stable/kinds_of_types.html#generators)

Thank you for linking this. I was thinking type checkers in general but looks like this is allowable across the board. We can keep the Iterator annotation then if we decide to keep this method as generator. I'm not really into statically typed Python if you can't tell. I mostly just use it for documentation of arguments and return types. 

> I don't see where's the complexity of using a generator?

I don't exactly see the benefit of changing how the NFS mount look ups to be a generator is a major improvement over just refactoring how the regex is used in the original implementation. Lazy loading can be nice, but the data we're working with here is simple enough that there's minimal, if any, benefit. I see 4 levels of nesting in the `_mounts(...)` function which starts to make me think there's potentially a more straightforward way to do things. That's my main issue. I also find that generators are more advanced and not everyone is familiar with them - especially folks where programming is not their primary role like admins - so it's my personal preference to only use a generator if I have a clear reason to. But that's my preference, not technical reasoning.

The regexps that we have are simple enough that I don't see any concern with maintaining them. Complex regexs can be cryptic - _[just look at the regex for finding an allowable password](https://ihateregex.io/expr/password)_ - but the two regexes here are pretty straight forward for filtering _/proc/mounts_. I don't think we'd be hurting from a maintenance perspective.

If we needed to get more granular in our fetching like finding mounts with specific mount options, then I can see the argument for using a generator and not the regexes. However, given that we're working with a pretty simple file, I don't think we need to take a hard line on whether to use a regex instead of a generator and vice versa in this specific case. We're not spending enough time in charm execution for it to really matter how we are processing the file. This method will really only be called when (a) the charm is mounting the NFS share, and (b) when the charm is unmounting the share when it is being destroyed by Juju.

So how about this? We don't use the regexes, but we modify the `_mounts` function just a bit?

```python3
def _mounts(fstype) -> Iterator[MountInfo]:
    with pathlib.Path("/proc/mounts").open(mode="rt") as mounts:
        for mount in mounts:
            # Lines in /proc/mounts follow a standard format.
            m = MountInfo(**mount.split())
            if not m.fstype.startswith(fstype):
                # Jump to next iteration
                continue

            yield m
```

A bit of inversion to make it a bit more clear what the "happy path" is. All I'd say is that we should probably also look at potentially taking out the URL check from the client so that we remove our reliance on the `re` module entirely in the client charm. The URL checker can be reimplemented in the nfs-server-proxy instead since that's where we really need to lint user input.

Reply:
Yep, pretty happy with your proposed change, since it's always better to reduce nesting on the function.